### ISSUE TITLE
Certain order of union members causes Incompatible types error

### ISSUE THREAD

**Issue by Stiivi** [python/mypy#3085]
I’m getting this weird error: Incompatible types in assignment (expression has type Dict[str, Union[str, float]], variable has type Dict[str, Union[str, int, float, bool]])

```python
Value = Union[str, int, float, bool]
d: Dict[str, Value]
d = {}
```

however this works:
```python
Value = Union[str, float, bool, int]
```

The order of union elements is different.

Other offenses:

`Value = Union[int, str, bool, float]`
`Value = Union[str, int, bool, float]`

Causes: `Incompatible types in assignment (expression has type Dict[str, Union[str, float]], variable has type Dict[str, Union[int, str, bool, float]])`

**Comment by gvanrossum** [comment]
Might have been fixed by #3025, which got rolled back for other reasons.

**Comment by Stiivi** [comment]
Another observation that looks related is this incorrect error message for the last statement of the following example. The last statement should cause _Incompatible types_ and it does:

```python
from typing import Union, Dict, Optional

Value = Union[str, float, bool, int]
d: Dict[str, Value]
d = {}

v: Value
v = d.get("athing")
```

However the error is showing wrong types:

```
Incompatible types in assignment (expression has type "Union[str, float, bool, None]",
                                    variable has type "Union[str, float, bool, int]")
```
 (indented for better demonstration)

The first union (expression type) in the error should have five elements – it is missing `int`.

**Comment by JelleZijlstra** [comment]
That's probably also something deciding that int is compatible with float, so `Union[float, int]` == float.

**Comment by ilevkivskyi** [comment]
I think this is essentially a duplicate of https://github.com/python/mypy/issues/1850

**Comment by JukkaL** [comment]
This should be fixed by #3025 (I'll need to double check) and yes, this looks like a duplicate of #1850 so I'm closing this now.
